### PR TITLE
python3Packages.biopython: patch SeqXMLIO parser, refactor

### LINKED ISSUE #292138: Build Failure: `biopython`
### Steps To Reproduce

Minimal flake to reproduce the issue:

`flake.nix`:

```nix
{
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";

  outputs = {
    self,
    nixpkgs,
  }: let
    forAllSystems = nixpkgs.lib.genAttrs [
      "aarch64-linux"
      "aarch64-darwin"
      "x86_64-darwin"
      "x86_64-linux"
    ];
    mkPkgs = system:
      import nixpkgs {
        inherit system;
        overlays = [
          # Uncomment this overlay and the issue will be fixed
          # (final: prev: {
          #   python3 = prev.python3.override {
          #     packageOverrides = pyfinal: pyprev: {
          #       biopython = pyprev.biopython.overrideAttrs (_: {
          #         doInstallCheck = false;
          #       });
          #     };
          #   };
          # })
        ];
      };
  in {
    # Development environment
    # Acessible through `nix develop`
    devShells = forAllSystems (system: let
      pkgs = mkPkgs system;
    in {
      default = pkgs.mkShell {
        nativeBuildInputs = [(pkgs.python3.withPackages (ps: [ps.biopython]))];
      };
    });
  };
}
```

`flake.lock`:

```nix
{
  "nodes": {
    "nixpkgs": {
      "locked": {
        "lastModified": 1708976803,
        "narHash": "sha256-yvRygcySjjSvj5JTaCdo7lPqJ/2mBV2XQ94Oaq/14qw=",
        "owner": "NixOS",
        "repo": "nixpkgs",
        "rev": "548a86b335d7ecd8b57ec617781f5e652ab0c38e",
        "type": "github"
      },
      "original": {
        "owner": "NixOS",
        "ref": "nixpkgs-unstable",
        "repo": "nixpkgs",
        "type": "github"
      }
    },
    "root": {
      "inputs": {
        "nixpkgs": "nixpkgs"
      }
    }
  },
  "root": "root",
  "version": 7
}
```

### Build log

[fail.log](https://github.com/NixOS/nixpkgs/files/14438672/fail.log)

### Additional context

Disabling install checks using `doInstallCheck = false` fixes the issue. If you uncomment the commented overlay in the minimal flake I provided, the build will succeed and the package will work properly.

### Notify maintainers

@luispedro 

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.74, NixOS, 24.05 (Uakari), 24.05.20240125.ae5c332`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/store/9s5qs4hni9fj88x79iw6im7amv7ghb76-source`
```

### PR DESCRIPTION
**Pull request by getchoo**
## Description of changes

Fixes #292138
ZHF #309482

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/biopython/default.nix [hunk 1]
```diff
@@ -18,15 +19,17 @@ buildPythonPackage rec {
     hash = "sha256-eOa/t43mMDQDev01/nfLbgqeW2Jwa+z3in2SKxbtg/c=";
   };
 
-  nativeBuildInputs = [
-    setuptools
+  patches = [
+    # cherry-picked from https://github.com/biopython/biopython/commit/3f9bda7ef44f533dadbaa0de29ac21929bc0b2f1
+    # fixes `IndexError: list index out of range` errors in test_embl7 & test_genbank8
```
Comment (annaleeleaves) [inline pkgs/development/python-modules/biopython/default.nix:24]:
```suggestion
    # fixes SeqXMLIO parser to process all data. remove on next update
```

**Review by ghost — APPROVED** [review]
Nice find! Change looks good. minor nits on patch. could copy bellow diff and just exclude first chunk tho
https://github.com/biopython/biopython/commit/3f9bda7ef44f533dadbaa0de29ac21929bc0b2f1.diff

`python3Packages` will not trigger `ofborg` builds, so python commits should be named `python311Packages.biopython: xxx`.

**Comment by annaleeleaves** [comment]
@ofborg build python31{1,2}Packages.biopython

**Review by ghost — APPROVED** [review]
Result of `nixpkgs-review pr 310411` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>eggnog-mapper</li>
    <li>eggnog-mapper.dist</li>
    <li>python311Packages.biopython</li>
    <li>python311Packages.biopython.dist</li>
    <li>python311Packages.cnvkit</li>
    <li>python311Packages.cnvkit.dist</li>
    <li>python311Packages.dnachisel</li>
    <li>python311Packages.dnachisel.dist</li>
    <li>python311Packages.genome-collector</li>
    <li>python311Packages.genome-collector.dist</li>
    <li>python312Packages.biopython</li>
    <li>python312Packages.biopython.dist</li>
    <li>python312Packages.dnachisel</li>
    <li>python312Packages.dnachisel.dist</li>
    <li>python312Packages.genome-collector</li>
    <li>python312Packages.genome-collector.dist</li>
  </ul>
</details>

**Pull request merged**